codalab/worker/local_run/docker_image_manager.py

```diff
@@ -143,82 +143,60 @@ def get(self, image_spec):
             image_spec += ':latest'
         try:
-            image = self._docker.images.get(image_spec)
-            digests = image.attrs.get('RepoDigests', [image_spec])
-            if len(digests) == 0:
+            if image_spec in self._downloading:
+                with self._downloading[image_spec]['lock']:
+                    if self._downloading[image_spec].is_alive():
+                        return ImageAvailabilityState(
+                            digest=None,
+                            stage=DependencyStage.DOWNLOADING,
+                            message=self._downloading[image_spec]['status'],
+                        )
+                    else:
+                        if self._downloading[image_spec]['success']:
+                            image = self._docker.images.get(image_spec)
+                            digest = image.attrs.get('RepoDigests', [image_spec])[0]
+                            with self._lock:
+                                self._image_cache[digest] = ImageCacheEntry(
+                                    id=image.id,
+                                    digest=digest,
+                                    last_used=time.time(),
+                                    virtual_size=image.attrs['VirtualSize'],
+                                    marginal_size=image.attrs['Size'],
+                                )
+                            status = ImageAvailabilityState(
+                                digest=digest, stage=DependencyStage.READY, message='Image ready'
+                            )
+                        else:
+                            status = ImageAvailabilityState(
+                                digest=None,
+                                stage=DependencyStage.FAILED,
+                                message=self._downloading[image_spec]['message'],
+                            )
+                        self._downloading.remove(image_spec)
+                        return status
+            else:
+
+                def download():
+                    logger.debug('Downloading Docker image %s', image_spec)
+                    try:
+                        self._docker.images.pull(image_spec)
+                        logger.debug('Download for Docker image %s complete', image_spec)
+                        self._downloading[image_spec]['success'] = True
+                        self._downloading[image_spec]['message'] = "Downloading image"
+                    except (docker.errors.APIError, docker.errors.ImageNotFound) as ex:
+                        logger.debug('Download for Docker image %s failed: %s', image_spec, ex)
+                        self._downloading[image_spec]['success'] = False
+                        self._downloading[image_spec][
+                            'message'
+                        ] = "Can't download image: {}".format(ex)
+
+                self._downloading.add_if_new(image_spec, threading.Thread(target=download, args=[]))
```

Comment:
Something seems weird here - `download` already sets `self._downloading`, so what is `add_if_new` doing on `self._downloading`?  Can you explain this more?

Reply:
I think `add_if_new` is to add `image_spec` to the dictionary but without any fields (`success` or `message`) being set. By the time when the `download` function gets executed and `pull` function finished, `self._downloading` will be able to set all the fields (`success`, `message`) for `image_spec`.

Reply:
I think add_if_new is to add image_spec to the dictionary but without any fields (success or message) being set. By the time when the download function gets executed and pull function finished, self._downloading will be able to set all the fields (success, message) for image_spec.

Reply:
`self._downloading` is actually a `ThreadDict` which is a convenience class I wrote ([docs here](https://docs.google.com/document/d/1BaclAazh-XmPfzIwLKCUMJTFKNiV6YrcDsIbUe_VMgk/edit#heading=h.sfbdizxrwv8n)). It basically lets us package threads with metadata keys in the form of a nested dict, so the thread itself is a value of the dict, alongside other key-value pairs (like `success` or `message`)

Like Jing said, `add_if_new`:
1. checks if we already have a download thread for this image spec (does nothing if that's the case)
2. creates and starts the download thread for this image spec
3. stores it as a value in the `ThreadDict`.

The `download` function itself only modifies other associated fields of the dict.